lib/AST/ASTStructuralEquivalence.cpp

```diff
@@ -102,11 +102,103 @@ static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
 static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
                                      const TemplateArgument &Arg1,
                                      const TemplateArgument &Arg2);
+static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
+                                     const TemplateArgumentList &ArgL1,
+                                     const TemplateArgumentList &ArgL2);
 static bool IsStructurallyEquivalent(StructuralEquivalenceContext &Context,
                                      NestedNameSpecifier *NNS1,
                                      NestedNameSpecifier *NNS2);
 static bool IsStructurallyEquivalent(const IdentifierInfo *Name1,
                                      const IdentifierInfo *Name2);
+static bool IsStructurallyEquivalent(const DeclarationName Name1,
+                                     const DeclarationName Name2);
+
+using ContextVector = llvm::SmallVector<const DeclContext *, 4>;
+
+static void GetContexts(ContextVector &Contexts, const NamedDecl *ND) {
+  const DeclContext *Ctx = ND->getDeclContext();
+
+  // For ObjC methods, look through categories and use the interface as context.
+  if (auto *MD = dyn_cast<ObjCMethodDecl>(ND))
+    if (auto *ID = MD->getClassInterface())
+      Ctx = ID;
+
+  while (Ctx && Ctx->isFunctionOrMethod())
```

Comment:
I am not sure that we can skip functions. Think about this case:
```
namespace A {
  void foo() {
    struct B {};
  }
  void bar() {
    struct B { int a; };
  }
}
```
Here the first and last `B` have a different scope.

Strictly speaking, the equivalency check should handle these cases too, shouldn't it?

Reply:
Yes theoretically it is possible to make such a structural equivalence check. But can this case occur from `ASTImporter`? During AST import of `bar` it is not possible to find (and compare) the `B` in `foo`? A `DeclContext` outside of `foo` should not have references to things inside `foo`, or can something "leak" to outside? (A special case can be the "struct definition in function parameter list" case.)

Reply:
Ok, you convinced me, let's skip the functions.